ocsp/responder/multi_source.go

```diff
@@ -54,50 +66,90 @@ func (src *multiSource) Response(ctx context.Context, req *ocsp.Request) (*Respo
 	// passes these checks, return its response instead.
 	select {
 	case <-ctx.Done():
-		src.counter.WithLabelValues("timed_out").Inc()
-		return nil, fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+		src.counter.WithLabelValues("primary_timed_out").Inc()
+		return nil, ctx.Err()
+
+	case r := <-primaryChan:
+		// If there was an error requesting from the primary, don't bother
+		// waiting for the secondary, because we wouldn't be able to
+		// check the secondary's status against the (more reliable) primary's
+		// status.
+		if r.err != nil {
+			src.counter.WithLabelValues("primary_error").Inc()
+			return nil, r.err
+		}
+		primaryResponse = r.resp
+	}
 
-	case primaryResult := <-primaryChan:
+	// The primary response was fresh enough to serve, go ahead and serve it.
+	if time.Since(primaryResponse.ThisUpdate) < src.expectedFreshness {
+		src.checkSecondary(primaryResponse, secondaryChan)
 		src.counter.WithLabelValues("primary_result").Inc()
-		return primaryResult.resp, primaryResult.err
+		return primaryResponse, nil
+	}
+
+	// The primary response was too stale to (ideally) serve. This will be
+	// a common path once we stop ocsp-updater from writing updated blobs
+	// to MariaDB. Try to serve from the secondary.
+	var secondaryResponse *Response
+	select {
+	case <-ctx.Done():
+		src.counter.WithLabelValues("timed_out_awaiting_secondary").Inc()
+		// Best-effort: return the primary response even though it's stale.
+		return primaryResponse, nil
 
 	case secondaryResult := <-secondaryChan:
-		// If secondary returns first, wait for primary to return for
-		// comparison.
-		var primaryResult responseResult
+		if secondaryResult.err != nil {
+			if errors.Is(secondaryResult.err, rocsp.ErrRedisNotFound) {
+				// This case will happen for several hours after first issuance.
+				src.counter.WithLabelValues("primary_stale_secondary_not_found").Inc()
+			} else {
+				src.counter.WithLabelValues("primary_stale_secondary_error").Inc()
+			}
+
+			// Best-effort: return the primary response even though it's stale.
+			return primaryResponse, nil
+		}
+		secondaryResponse = secondaryResult.resp
+	}
 
-		// Listen for cancellation or timeout waiting for primary result.
-		select {
-		case <-ctx.Done():
-			src.counter.WithLabelValues("timed_out").Inc()
-			return nil, fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+	// If the secondary response status doesn't match primary, return
+	// primary response. For instance this will happen for several hours
+	// after any revocation.
+	if secondaryResponse.Status != primaryResponse.Status {
+		src.counter.WithLabelValues("stale_primary_status_wins").Inc()
```

Comment:
nit: I think this is the only label value that doesn't start with "primary_" or "secondary_"; consider rephrasing.